pam/integration-tests/gdm-integration_test.go

```diff
@@ -0,0 +1,389 @@
+		},
+		"Error on authenticating user2 with too many retries": {
+			pamUser: "user2",
+			authModeIDs: []string{
+				passwordAuthID,
+				passwordAuthID,
+				passwordAuthID,
+				passwordAuthID,
+				passwordAuthID,
+				passwordAuthID,
+			},
+			eventPollResponses: map[gdm.EventType][]*gdm.EventData{
+				gdm.EventType_startAuthentication: {
+					gdm_test.IsAuthenticatedEvent(&authd.IARequest_AuthenticationData_Challenge{
+						Challenge: "not goodpass",
+					}),
+					gdm_test.IsAuthenticatedEvent(&authd.IARequest_AuthenticationData_Challenge{
+						Challenge: "another not goodpass",
+					}),
+					gdm_test.IsAuthenticatedEvent(&authd.IARequest_AuthenticationData_Challenge{
+						Challenge: "even more not goodpass",
+					}),
+					gdm_test.IsAuthenticatedEvent(&authd.IARequest_AuthenticationData_Challenge{
+						Challenge: "not yet goodpass",
+					}),
+					gdm_test.IsAuthenticatedEvent(&authd.IARequest_AuthenticationData_Challenge{
+						Challenge: "really, it's not a goodpass!",
+					}),
+					gdm_test.IsAuthenticatedEvent(&authd.IARequest_AuthenticationData_Challenge{
+						Challenge: "goodpass",
+					}),
+				},
+			},
+			wantPamErrorMessages: []string{
+				"invalid password 'really, it's not a goodpass!', should be 'goodpass'",
+			},
+			wantError:       pam.ErrAuth,
+			wantAcctMgmtErr: ignoreError,
+		},
+		"Error on authenticating unknown user": {
+			pamUser: "user-unknown",
+			eventPollResponses: map[gdm.EventType][]*gdm.EventData{
+				gdm.EventType_startAuthentication: {
+					gdm_test.IsAuthenticatedEvent(&authd.IARequest_AuthenticationData_Challenge{
+						Challenge: "",
+					}),
+				},
+			},
+			wantPamErrorMessages: []string{
+				"user not found",
+			},
+			wantError:       pam.ErrAuth,
+			wantAcctMgmtErr: ignoreError,
+		},
+		"Error on invalid fido ack": {
+			pamUser:     "user-mfa",
+			authModeIDs: []string{passwordAuthID, fido1AuthID},
+			eventPollResponses: map[gdm.EventType][]*gdm.EventData{
+				gdm.EventType_startAuthentication: {
+					gdm_test.IsAuthenticatedEvent(&authd.IARequest_AuthenticationData_Challenge{
+						Challenge: "goodpass",
+					}),
+					gdm_test.IsAuthenticatedEvent(&authd.IARequest_AuthenticationData_Wait{}),
+				},
+			},
+			wantPamErrorMessages: []string{
+				fido1AuthID + " should have wait set to true",
+			},
+			wantError:       pam.ErrAuth,
+			wantAcctMgmtErr: ignoreError,
+		},
+	}
+	for name, tc := range testCases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Cleanup(pam_test.MaybeDoLeakCheck)
+			t.Parallel()
+
+			// We run a daemon for each test, because here we don't want to
+			// make assumptions whether the state of the broker and each test
+			// should run in parallel and work the same way in any order is ran.
+			ctx, cancel := context.WithCancel(context.Background())
+			socketPath, stopped := testutils.RunDaemon(ctx, t, daemonPath,
+				testutils.WithEnvironment(grouptests.GPasswdMockEnv(t, gpasswdOutput, groupsFile)...),
+			)
```

Comment:
Do we need a daemon per test case? It should be fine running a single one for TestGdmModule, no?

Reply:
That was what I was doing initially, but then that implies that if there are concurrent tests, they may break each other depending on the order they're running.

Since we're doing assumptions about the daemon remembering things or the previously used auth method for user.

And if we really we want to do them, we should do them in another test. Here I cared more about the interaction with the model itself in a real scenario where we don't have more than one client logging through gdm anyways.